lib/api/model/model.dart

```diff
@@ -31,6 +31,92 @@ class CustomProfileField {
   Map<String, dynamic> toJson() => _$CustomProfileFieldToJson(this);
 }
 
+@JsonSerializable(fieldRename: FieldRename.snake)
+class ProfileFieldUserData {
+  final String value;
+  final String? renderedValue;
+
+  ProfileFieldUserData({required this.value, this.renderedValue});
+
+  factory ProfileFieldUserData.fromJson(Map<String, dynamic> json) =>
+    _$ProfileFieldUserDataFromJson(json);
+
+  Map<String, dynamic> toJson() => _$ProfileFieldUserDataToJson(this);
+}
+
+/// As in [InitialSnapshot.realmUsers], [InitialSnapshot.realmNonActiveUsers], and [InitialSnapshot.crossRealmBots].
+///
+/// In the Zulip API, the items in realm_users, realm_non_active_users, and
+/// cross_realm_bots are all extremely similar. They differ only in that
+/// cross_realm_bots has is_system_bot.
+///
+/// https://zulip.com/api/register-queue#response
+@JsonSerializable(fieldRename: FieldRename.snake)
+class User {
+  final int userId;
+  @JsonKey(name: 'delivery_email')
+  String? deliveryEmailStaleDoNotUse; // TODO see [RealmUserUpdateEvent.deliveryEmail]
+  String email;
+  String fullName;
+  String dateJoined;
+  bool isActive; // Really sometimes absent in /register, but we normalize that away; see [InitialSnapshot.realmUsers].
+  bool isOwner;
+  bool isAdmin;
+  bool isGuest;
+  bool? isBillingAdmin; // TODO(server-5)
+  bool isBot;
+  int? botType; // TODO enum
+  int? botOwnerId;
+  int role; // TODO enum
+  String timezone;
+  String? avatarUrl; // TODO distinguish null from missing https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/flutter.3A.20omitted.20vs.2E.20null.20in.20JSON/near/1551759
+  int avatarVersion;
+  // null for bots, which don't have custom profile fields.
+  // If null for a non-bot, equivalent to `{}` (null just written for efficiency.)
+  @JsonKey(readValue: _readProfileData)
+  Map<int, ProfileFieldUserData>? profileData;
+  @JsonKey(readValue: _readIsSystemBot)
+  bool? isSystemBot; // TODO(server-5)
+
+  static bool? _readIsSystemBot(Map json, String key) =>
+    json[key] ?? json['is_cross_realm_bot'];
+
+  static Map<int, ProfileFieldUserData>? _readProfileData(json, key) {
+    final value = (json[key] as Map<int, ProfileFieldUserData>?);
```

Comment:
Hmm, this doesn't seem right. See this in [the doc](https://pub.dev/documentation/json_annotation/latest/json_annotation/JsonKey/readValue.html):
> Note: using this feature does not change any of the subsequent decoding logic for the field. For instance, if the field is of type [DateTime](https://api.dart.dev/stable/2.19.6/dart-core/DateTime-class.html) we expect the function provided here to return a [String](https://api.dart.dev/stable/2.19.6/dart-core/String-class.html).

That's a bit opaque; but more concretely, see the generated code (which this revision is missing, btw):
```dart
      profileData:
          (User._readProfileData(json, 'profile_data') as Map<String, dynamic>?)
              ?.map(
        (k, e) => MapEntry(int.parse(k),
            ProfileFieldUserData.fromJson(e as Map<String, dynamic>)),
      ),
```

That also reminds me of something I should have thought of earlier: we should have some unit tests for the nontrivial logic in our JSON decoding. No need to have it for all the boring fields — I don't think tests for those would do much good — but for basically the places we're using `readValue`.